### PR TITLE
Dsindex docstrings

### PR DESCRIPTION
**Pull request by SergeyTsimfer**

### CODE REVIEW

File: batchflow/dsindex.py [hunk 1]
```diff
@@ -34,12 +39,48 @@ def from_index(cls, *args, **kwargs):
...
+
+        Parameters
+        ----------
+        index_list : list
+            indices to be concatenated. each item in the list should
+            have method .index, that returns an 1-d array-like structure
+
+        Returns
+        -------
+        DatasetIndex with one common index
```
Comment (roman-kh) [inline batchflow/dsindex.py]: Check numpydoc format: data type and a decsription should be stated on separate lines

File: batchflow/dsindex.py [hunk 2]
```diff
@@ -34,12 +39,48 @@ def from_index(cls, *args, **kwargs):
...
     @classmethod
     def concat(cls, *index_list):
-        """ Create index by concatenating other indices. """
+        """Create index by concatenating other indices.
+
+        Parameters
+        ----------
+        index_list : list
+            indices to be concatenated. each item in the list should
+            have method .index, that returns an 1-d array-like structure
```
Comment (roman-kh) [inline batchflow/dsindex.py]: Each item is expected to be an instance of DatasetIndex or contain ``index`` property with 1d sequence of indices

File: batchflow/dsindex.py [hunk 3]
```diff
@@ -34,12 +39,48 @@ def from_index(cls, *args, **kwargs):
...
+        index : int, 1-d array-like or callable
+            defines structure of DatasetIndex
+
+            - list, numpy.array, pandas.DataSeries - structure is np.array
+
+            - int - structure is np.arange() of given length
+
+            - callable - a function which returns data that can
+                be represented as array
+
```
Comment (roman-kh) [inline batchflow/dsindex.py]: One can hardly understand what this text means. Try to reformulate it more clearly

File: batchflow/dsindex.py [hunk 4]
```diff
@@ -34,12 +39,48 @@ def from_index(cls, *args, **kwargs):
...
+        ------
+        TypeError
+            if 'index' is not 1-dimensional
+
+        ValueError
+            if 'index' is empty
+
+        Returns
+        -------
+            index to be stored in class instance
```
Comment (roman-kh) [inline batchflow/dsindex.py]: data type?

File: batchflow/dsindex.py [hunk 5]
```diff
@@ -68,7 +109,30 @@ def build_pos(self):
...
+            items to return position of
+
+            - int, str - return position of that item in the DatasetIndex
+
+            - slice, Iterable - return positions of multiple items,
+                specified by argument
+
+        Returns
+        -------
+        Positions in DatasetIndex of specified items
```
Comment (roman-kh) [inline batchflow/dsindex.py]: positions of ... in ...

File: batchflow/dsindex.py [hunk 6]
```diff
@@ -68,7 +109,30 @@ def build_pos(self):
...
+
+        Examples
+        --------
+
```
Comment (roman-kh) [inline batchflow/dsindex.py]: empty line is not required here

File: batchflow/dsindex.py [hunk 7]
```diff
@@ -347,9 +422,36 @@ def gen_batch(self, batch_size, shuffle=False, n_epochs=1, drop_last=False, bar=
...
     def create_batch(self, batch_indices, pos=True, as_array=False, *args, **kwargs):
         """ Create a batch from given indices.
-        if `pos` is `False`, then `batch_indices` should contain the indices
-        which should be included in the batch (so expected batch is just the very same batch_indices)
-        otherwise `batch_indices` contains positions in the current index.
+
+        Parameters
+        ----------
+
+        batch_indices : int, slice, list, numpy.array or DatasetIndex
```
Comment (roman-kh) [inline batchflow/dsindex.py]: Rename ``batch_indices`` to ``index``

File: batchflow/dsindex.py [hunk 8]
```diff
@@ -347,9 +422,36 @@ def gen_batch(self, batch_size, shuffle=False, n_epochs=1, drop_last=False, bar=
...
+            indices to be returned as separate batch
+            (so expected batch is just the very same batch_indices)
+
+        pos : bool - flag that determines how function works
+
+        as_array : bool - flag that determines type of returned value
+
+        Returns
+        -------
+        batch : DatasetIndex or numpy.array
```
Comment (roman-kh) [inline batchflow/dsindex.py]: No need to state a variable name, just its data type

File: batchflow/dsindex.py [hunk 9]
```diff
@@ -347,9 +422,36 @@ def gen_batch(self, batch_size, shuffle=False, n_epochs=1, drop_last=False, bar=
...
+            part of initial DatasetIndex, specified by 'batch_indices'
+
+        Examples
+        --------
+
```
Comment (roman-kh) [inline batchflow/dsindex.py]: Unnecessary empty line

File: batchflow/dsindex.py [hunk 10]
```diff
@@ -82,7 +146,18 @@ def get_pos(self, index):
...
+        Parameters
+        ----------
+
+        pos : int, slice, list or numpy.array
+            positions of items to include in subset
+
+        Returns
+        -------
+        numpy.array - subset of DatasetIndex.index
```
Comment (AlexeyKozhevin) [inline batchflow/dsindex.py]: Check doc-style:
```
name : type
    comment
```
or
```
type
    comment
```

File: batchflow/dsindex.py [hunk 11]
```diff
@@ -347,9 +422,36 @@ def gen_batch(self, batch_size, shuffle=False, n_epochs=1, drop_last=False, bar=
...
+        batch : DatasetIndex or numpy.array
+            part of initial DatasetIndex, specified by 'batch_indices'
+
+        Examples
+        --------
+
+        Create DatasetIndex with first 100 natural numbers, then
+        get batch with every second item
+
+        >>>DatasetIndex(100).create_batch(batch_indices=2*np.arange(50))
```
Comment (AlexeyKozhevin) [inline batchflow/dsindex.py]: add whitespace after `>>>`

File: batchflow/dsindex.py [hunk 12]
```diff
@@ -347,9 +422,36 @@ def gen_batch(self, batch_size, shuffle=False, n_epochs=1, drop_last=False, bar=
...
+        batch_indices : int, slice, list, numpy.array or DatasetIndex
+            if 'pos' is True, then 'batch_indices' should contain
+            positions of items in the current index to be returned as
+            separate batch
+
+            if 'pos' is False, then 'batch_indices' should contain
+            indices to be returned as separate batch
+            (so expected batch is just the very same batch_indices)
+
+        pos : bool - flag that determines how function works
```
Comment (AlexeyKozhevin) [inline batchflow/dsindex.py]: Check docstyle

File: batchflow/dsindex.py [hunk 13]
```diff
@@ -100,7 +175,7 @@ def split(self, shares=0.8, shuffle=False):
         ----------
         shares : float or tuple of floats - train, test and validation shares.
 
-        shuffle : bool - whether to shuffle the index before split.
+        shuffle : bool, int or callable - whether to shuffle the index before split.
```
Comment (AlexeyKozhevin) [inline batchflow/dsindex.py]: Check docstyle everywhere

**Comment by roman-kh** [comment]
Make sure that test code is checked by pylint

File: batchflow/dsindex.py [hunk 14]
```diff
@@ -14,13 +14,18 @@ class DatasetIndex(Baseset):
...
     Examples
     --------
-    >>> index = DatasetIndex(all_item_ids)
+    >>> ds_index = DatasetIndex(all_item_ids)
 
-    >>> index.split([0.8, 0.2])
+    >>> ds_index.split([0.8, 0.2])
 
-    >>> item_pos = index.get_pos(item_id)
+    >>> item_pos = ds_index.get_pos(item_id)
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply batchflow/dsindex.py]: Tried to add some clarity and distinction between `index`(structure), `DatasetIndex`(class that provides additional operations on structure such as `create_batch`

File: batchflow/dsindex.py [hunk 15]
```diff
@@ -60,26 +60,29 @@ def build_index(index):
         Parameters
         ----------
         index : int, 1-d array-like or callable
-            defines structure of DatasetIndex
+            Defines structure of DatasetIndex
```
Comment (roman-kh) [inline batchflow/dsindex.py]: I would say this defines the content rather than the structure

File: batchflow/dsindex.py [hunk 16]
```diff
@@ -60,26 +60,29 @@ def build_index(index):
...
 
-            - list, numpy.array, pandas.DataSeries - structure is np.array
+            - list, numpy.array, pandas.DataSeries
```
Comment (roman-kh) [inline batchflow/dsindex.py]: just any array-like

**Comment by roman-kh** [comment]
@AlexeyKozhevin, please, check the PR

**Comment by SergeyTsimfer** [comment]
Branch was merged to dsindex tests

**Pull request closed without merging**